src/ucp/core/ucp_mm.inl

```diff
@@ -73,39 +73,62 @@ not_found:
 }
 
 static UCS_F_ALWAYS_INLINE void
-ucp_memh_put(ucp_context_h context, ucp_mem_h memh)
+ucp_memh_invalidate(ucp_context_h context, ucp_mem_h memh,
+                    ucs_rcache_invalidate_comp_func_t cb, void *arg,
+                    ucp_md_map_t inv_md_map)
 {
-    ucs_rcache_t *rcache;
-    khiter_t iter;
-
-    ucs_trace("memh %p: release address %p length %zu md_map %" PRIx64,
+    ucs_trace("memh %p: release address %p length %zu md_map %" PRIx64
+              " inv_md_map %" PRIx64,
               memh, ucp_memh_address(memh), ucp_memh_length(memh),
-              memh->md_map);
+              memh->md_map, inv_md_map);
 
     if (ucs_unlikely(ucp_memh_is_zero_length(memh))) {
         return;
     }
 
+    memh->inv_md_map |= inv_md_map;
+
     /* User memory handle, or rcache was disabled */
     if (ucs_unlikely(memh->parent != NULL)) {
-        ucp_memh_cleanup(context, memh);
-        ucs_free(memh);
+        if (!--memh->super.refcount) {
+            ucp_memh_cleanup(context, memh);
+            ucs_free(memh);
+
+            if (cb != NULL) {
+                cb(arg);
+            }
+        }
         return;
     }
 
     UCP_THREAD_CS_ENTER(&context->mt_lock);
+    if (cb != NULL) {
```

Comment:
maybe check inv_md_map?

Reply:
we may want invalidate with inv_md_map 0